components/scheduler/src/main/java/org/cardanofoundation/ledgersync/scheduler/service/offchain/OffChainFetchAbstractService.java

```diff
@@ -0,0 +1,239 @@
+import com.fasterxml.jackson.databind.DeserializationFeature;
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.netty.channel.ChannelOption;
+import io.netty.handler.ssl.SslContextBuilder;
+import io.netty.handler.ssl.SslProvider;
+import io.netty.handler.ssl.util.InsecureTrustManagerFactory;
+import io.netty.handler.timeout.ReadTimeoutHandler;
+import io.netty.handler.timeout.WriteTimeoutHandler;
+import jakarta.annotation.PostConstruct;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import javax.net.ssl.SSLException;
+import lombok.AccessLevel;
+import lombok.experimental.FieldDefaults;
+import lombok.extern.slf4j.Slf4j;
+import org.cardanofoundation.ledgersync.common.util.UrlUtil;
+import org.cardanofoundation.ledgersync.scheduler.dto.anchor.AnchorDTO;
+import org.cardanofoundation.ledgersync.scheduler.dto.offchain.OffChainFetchResultDTO;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.HttpStatusCode;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.http.client.reactive.ReactorClientHttpConnector;
+import org.springframework.web.reactive.function.client.WebClient;
+import reactor.netty.http.client.HttpClient;
+
+
+@FieldDefaults(level = AccessLevel.PRIVATE)
+@Slf4j
+public abstract class OffChainFetchAbstractService<S, F, O extends OffChainFetchResultDTO, T extends AnchorDTO> {
+
+    static final int TIMEOUT = 30000;
+    static final int READ_TIMEOUT = 19000;
+    static final int WRITE_TIMEOUT = 10000;
+    static final int LIMIT_BYTES = 4096;
+
+    protected Queue<O> offChainAnchorsFetchResult;
+
+    public void initOffChainListData() {
+        offChainAnchorsFetchResult = new ConcurrentLinkedQueue<>();
+    }
+
+    public abstract S extractOffChainData(O offChainAnchorData, Integer maxRetry);
+
+    public abstract F extractFetchError(O offChainAnchorData);
+
+    public abstract O addIdKey(OffChainFetchResultDTO offChainAnchorData, T anchor);
+
+    final ObjectMapper objectMapper = new ObjectMapper();
+
+    @PostConstruct
+    void setup() {
+        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
+    }
+
+    public List<S> getOffChainAnchorsFetch(Integer maxRetry) {
+        return offChainAnchorsFetchResult.stream()
+            .map(e -> this.extractOffChainData(e, maxRetry))
```

Comment:
It looks like you're using maxRetry here just for the setter when extracting data. I think you could remove the parameter from the method and use maxRetry externally